cpp/src/gandiva/function_registry.h

```diff
@@ -52,9 +58,24 @@ class GANDIVA_EXPORT FunctionRegistry {
   arrow::Status Register(const std::vector<NativeFunction>& funcs,
                          std::shared_ptr<arrow::Buffer> bitcode_buffer);
 
+  /// \brief register a stub function into the function registry
+  /// @param func the registered function's metadata
+  /// @param stub_function_ptr the function pointer to the
+  /// registered function's implementation
+  /// @param function_holder_maker this will be used as the function holder if the
+  /// function requires a function holder
+  arrow::Status Register(
+      NativeFunction func, void* stub_function_ptr,
+      std::optional<FunctionHolderMaker> function_holder_maker = std::nullopt);
```

Comment:
This is the new API used for registering external stub functions. If a function requires function holder, it can use the `function_holder_maker` to specify it